src/MatrixProductOperator.jl

```diff
@@ -0,0 +1,175 @@
+using OptimizedEinsum: ContractionPath
+using UUIDs: uuid4
+using IterTools: partition
+using Random
+
+abstract type MatrixProductOperator{B} <: State{B} end
+
+function MatrixProductOperator(arrays; bounds::Type{<:Bounds} = Open, kwargs...)
+    MatrixProductOperator{bounds}(arrays; kwargs...)
+end
+
+function MatrixProductOperator{Open}(arrays; χ = nothing, order = (:l, :r, :p), meta...)
+    !issetequal(order, (:l, :r, :p)) && throw(ArgumentError("`order` must be a permutation of the :l, :r and :p"))
+    order = Dict(dim => i for (i, dim) in enumerate(order))
+
+    # check format
+    # NOTE matching index dimensions are checked on `push!(::TensorNetwork,...)`
+    # TODO check χ?
+    if !all(==(4) ∘ ndims, arrays[2:end-1]) || ndims(first(arrays)) != 3 || ndims(last(arrays)) != 3
+        throw(DimensionMismatch("virtual bond-dims mismatch"))
+    end
+
+    tn = TensorNetwork{MatrixProductOperator{Open}}(; χ = χ, order = order, meta...)
+
+    n = length(arrays)
+    vinds = Dict(x => Symbol(uuid4()) for x in partition(1:n, 2, 1))
+    oinds = Dict(i => Symbol(uuid4()) for i in 1:n)
+    iinds = Dict(i => Symbol(uuid4()) for i in 1:n)
+    permutator = [order[i] for i in (:l, :r, :p)]
```

Comment:
Here also you need another physical index

Reply:
Thanks! Code is not ready yet.